Source/WebGPU/WGSL/TypeCheck.cpp

```diff
@@ -0,0 +1,321 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "TypeCheck.h"
+
+#include "AST.h"
+#include "ASTVisitor.h"
+#include "ContextProviderInlines.h"
+#include <wtf/SetForScope.h>
+#include <wtf/Vector.h>
+
+namespace WGSL {
+
+class UnificationContext;
+
+struct Type {
+    enum Kind {
+        Variable,
+        Primitive,
+        Constructor,
+        Application,
+        Function,
+    };
+};
+
+using SubstituionCallback = std::function<void(Type*)>;
+
+class TypeChecker : public AST::Visitor, public ContextProvider<Type*> {
+    friend class UnificationContext;
+
+public:
+    TypeChecker(AST::ShaderModule& shaderModule)
+        : m_shaderModule(shaderModule)
+    {
+    }
+
+    void check();
+
+    void visit(AST::Structure&) override;
+    void visit(AST::Variable&) override;
+    void visit(AST::Function&) override;
+    void visit(AST::Statement&) override;
+    void visit(AST::AssignmentStatement&) override;
+    void visit(AST::ReturnStatement&) override;
+    void visit(AST::Expression&) override;
+    void visit(AST::FieldAccessExpression&) override;
+    void visit(AST::IndexAccessExpression&) override;
+    void visit(AST::BinaryExpression&) override;
+    void visit(AST::IdentifierExpression&) override;
+    void visit(AST::TypeName&) override;
+
+private:
+    Type* infer(AST::Expression&);
+    Type* variable();
+    Type* resolve(AST::TypeName&);
+    void inferred(Type*);
+    void unify(Type*, Type*);
+
+    void substitute(Type*, SubstituionCallback&&);
+
+    AST::ShaderModule& m_shaderModule;
+    Type* m_inferredType { nullptr };
+    UnificationContext* m_unificationContext { nullptr };
+
+    // FIXME: remove this once we start allocating types
+    Type m_type;
+};
+
+class UnificationContext {
+public:
+    UnificationContext(TypeChecker* typeChecker)
+        : m_setForScope(typeChecker->m_unificationContext, this)
+    {
+    }
+
+    ~UnificationContext()
+    {
+        solveConstraints();
+        runSubstitutions();
+    }
+
+    void unify(Type* lhs, Type* rhs)
+    {
+        m_constraints.append({ lhs, rhs });
+    }
+
+    void substitute(Type* type, SubstituionCallback&& callback)
+    {
+        m_substitutions.append({ type, WTFMove(callback) });
+    }
+
+private:
+    void solveConstraints()
+    {
+        // FIXME: implement this
+    }
+
+    void runSubstitutions()
+    {
+        // FIXME: implement this
+    }
+
+    SetForScope<UnificationContext*> m_setForScope;
+    Vector<std::pair<Type*, Type*>> m_constraints;
+    Vector<std::pair<Type*, SubstituionCallback>> m_substitutions;
+};
+
+void TypeChecker::check()
+{
+    for (auto& structure : m_shaderModule.structures())
+        visit(structure);
+
+    for (auto& variable : m_shaderModule.variables())
+        visit(variable);
+
+    for (auto& function : m_shaderModule.functions())
+        visit(function);
```

Comment:
Visiting a function doesn't visit its body? That seems confusing...

Reply:
First we declare all the functions, then we visit the bodies.